src/main/java/ch/njol/skript/expressions/ExprMidpoint.java

```diff
@@ -0,0 +1,146 @@
+@Since("INSERT VERSION")
+public class ExprMidpoint extends SimpleExpression<Object> implements SyntaxRuntimeErrorProducer {
+
+	private static final int NONE = 0, LOCATION = 1, VECTOR = 2, BOTH = 3;
+
+	static {
+		Skript.registerExpression(ExprMidpoint.class, Object.class, ExpressionType.COMBINED,
+			"[the] mid[-]point (of|between) %location/vector% and %location/vector%");
+	}
+
+	private Expression<?> object1;
+	private Expression<?> object2;
+	private Class<?> classType = null;
+	private Node node;
+
+	@Override
+	public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelayed, ParseResult parseResult) {
+		object1 = exprs[0];
+		object2 = exprs[1];
+		int type1 = checkExpressionType(object1);
+		int type2 = checkExpressionType(object2);
+		if (type1 != type2) {
+			// If both 'type1' and 'type2' are either a Location or a Vector, then error, since they don't match
+			if ((type1 == LOCATION || type1 == VECTOR) && (type2 == LOCATION || type2 == VECTOR)) {
+				Skript.error("You can only get the midpoint between two locations or two vectors.");
+				return false;
+			}
+			// Otherwise, checks will be done in 'get'
+		} else if (type1 == LOCATION) {
+			// Both 'type1' and 'type2' are a Location
+			classType = Location.class;
+		} else if (type1 == VECTOR) {
+			// Both 'type1' and 'type2' are a Vector
+			classType = Vector.class;
+		}
+		// Otherwise, checks will be done in 'get'
+		node = getParser().getNode();
+		return true;
+	}
+
+	@Override
+	protected Object @Nullable [] get(Event event) {
+		Object object1 = this.object1.getSingle(event);
+		Object object2 = this.object2.getSingle(event);
+		if (object1 == null || object2 == null) {
+			return null;
+		} else if (classType == null) {
+			if (object1 instanceof Location && object2 instanceof Location) {
+				classType = Location.class;
+			} else if (object1 instanceof Vector && object2 instanceof Vector) {
+				classType = Vector.class;
+			} else {
+				error("You can only get the midpoint between two locations or two vectors.");
+				return null;
+			}
+		}
+		if (classType.equals(Location.class)) {
+			Location loc1 = (Location) object1;
+			Location loc2 = (Location) object2;
+			if (loc1.getWorld() != loc2.getWorld()) {
+				error("Cannot get the midpoint of two locations in different worlds.");
+				return null;
+			}
+			World world = loc1.getWorld();
+			Vector vector = loc1.toVector().getMidpoint(loc2.toVector());
+			return new Location[] {vector.toLocation(world)};
+		} else {
+			Vector vector1 = (Vector) object1;
+			Vector vector2 = (Vector) object2;
+			return new Vector[] {vector1.getMidpoint(vector2)};
+		}
+	}
+
+	private int checkExpressionType(Expression<?> expr) {
+		if (expr.canReturn(Location.class)) {
+			if (expr.canReturn(Vector.class))
+				return BOTH;
+			return LOCATION;
+		} else if (expr.canReturn(Vector.class)) {
+			return VECTOR;
+		}
+		return NONE;
+	}
```

Comment:
Like i mentioned above, it's probably better to have this return a class array of whatever it can return
then in init you can just compare the arrays (both length 1 and not equal? error) and otherwise just store the larger one.
Then possibleReturnTypes can just return the stored array and getReturnType can return the supertype (which can be computed at init)